Trying to access session_database.key before it is created

### Steps to reproduce

1. Fresh install on an Android 6 emulator
2. Launch the app


### Outcome

#### What did you expect?
App to start successfully
#### What happened instead?
App crashes immediately with the following exception
```
FATAL EXCEPTION: main
Process: io.element.android.x.debug, PID: 4020
java.lang.RuntimeException: Unable to create application io.element.android.x.ElementXApplication: androidx.startup.StartupException: java.io.FileNotFoundException: /data/user/0/io.element.android.x.debug/databases/session_database.key: open failed: ENOENT (No such file or directory)
	at android.app.ActivityThread.handleBindApplication(ActivityThread.java:4710)
	at android.app.ActivityThread.-wrap1(ActivityThread.java)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1405)
	at android.os.Handler.dispatchMessage(Handler.java:102)
	at android.os.Looper.loop(Looper.java:148)
	at android.app.ActivityThread.main(ActivityThread.java:5417)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
Caused by: androidx.startup.StartupException: java.io.FileNotFoundException: /data/user/0/io.element.android.x.debug/databases/session_database.key: open failed: ENOENT (No such file or directory)
	at androidx.startup.AppInitializer.doInitialize(AppInitializer.java:187)
	at androidx.startup.AppInitializer.doInitialize(AppInitializer.java:138)
	at androidx.startup.AppInitializer.initializeComponent(AppInitializer.java:117)
	at io.element.android.x.ElementXApplication.onCreate(ElementXApplication.kt:36)
	at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1013)
	at android.app.ActivityThread.handleBindApplication(ActivityThread.java:4707)
	at android.app.ActivityThread.-wrap1(ActivityThread.java) 
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1405) 
	at android.os.Handler.dispatchMessage(Handler.java:102) 
	at android.os.Looper.loop(Looper.java:148) 
	at android.app.ActivityThread.main(ActivityThread.java:5417) 
	at java.lang.reflect.Method.invoke(Native Method) 
	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726) 
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616) 
Caused by: java.io.FileNotFoundException: /data/user/0/io.element.android.x.debug/databases/session_database.key: open failed: ENOENT (No such file or directory)
	at libcore.io.IoBridge.open(IoBridge.java:452)
	at java.io.FileOutputStream.<init>(FileOutputStream.java:87)
	at java.io.FileOutputStream.<init>(FileOutputStream.java:72)
	at androidx.security.crypto.EncryptedFile.openFileOutput(EncryptedFile.java:262)
	at io.element.encrypteddb.passphrase.RandomSecretPassphraseProvider.getPassphrase(RandomSecretPassphraseProvider.kt:40)
	at io.element.encrypteddb.SqlCipherDriverFactory.create(SqlCipherDriverFactory.kt:41)
	at io.element.android.libraries.sessionstorage.impl.di.SessionStorageModule.provideMatrixDatabase(SessionStorageModule.kt:40)
	at io.element.android.libraries.sessionstorage.impl.di.SessionStorageModule_ProvideMatrixDatabaseFactory$Companion.provideMatrixDatabase(SessionStorageModule_ProvideMatrixDatabaseFactory.kt:32)
	at io.element.android.libraries.sessionstorage.impl.di.SessionStorageModule_ProvideMatrixDatabaseFactory.get(SessionStorageModule_ProvideMatrixDatabaseFactory.kt:23)
	at io.element.android.libraries.sessionstorage.impl.di.SessionStorageModule_ProvideMatrixDatabaseFactory.get(SessionStorageModule_ProvideMatrixDatabaseFactory.kt:20)
	at dagger.internal.DoubleCheck.get(DoubleCheck.java:47)
	at io.element.android.libraries.sessionstorage.impl.DatabaseSessionStore_Factory.get(DatabaseSessionStore_Factory.kt:21)
	at io.element.android.libraries.sessionstorage.impl.DatabaseSessionStore_Factory.get(DatabaseSessionStore_Factory.kt:17)
	at dagger.internal.DoubleCheck.get(DoubleCheck.java:47)
	at io.element.android.features.rageshake.impl.reporter.DefaultBugReporter_Factory.get(DefaultBugReporter_Factory.kt:40)
	at io.element.android.features.rageshake.impl.reporter.DefaultBugReporter_Factory.get(DefaultBugReporter_Factory.kt:26)
	at dagger.internal.DoubleCheck.get(DoubleCheck.java:47)
	at io.element.android.x.di.DaggerAppComponent$AppComponentImpl.bugReporter(DaggerAppComponent.java:2969)
	at io.element.android.x.initializer.TracingInitializer.create(TracingInitializer.kt:37)
	at io.element.android.x.initializer.TracingInitializer.create(TracingInitializer.kt:33)
	at androidx.startup.AppInitializer.doInitialize(AppInitializer.java:180)
	at androidx.startup.AppInitializer.doInitialize(AppInitializer.java:138) 
	at androidx.startup.AppInitializer.initializeComponent(AppInitializer.java:117) 
	at io.element.android.x.ElementXApplication.onCreate(ElementXApplication.kt:36) 
	at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1013) 
	at android.app.ActivityThread.handleBindApplication(ActivityThread.java:4707) 
	at android.app.ActivityThread.-wrap1(ActivityThread.java) 
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1405) 
	at android.os.Handler.dispatchMessage(Handler.java:102) 
	at android.os.Looper.loop(Looper.java:148) 
	at android.app.ActivityThread.main(ActivityThread.java:5417) 
	at java.lang.reflect.Method.invoke(Native Method) 
	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726) 
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616) 
Caused by: android.system.ErrnoException: open failed: ENOENT (No such file or directory)
	at libcore.io.Posix.open(Native Method)
	at libcore
```

### Your phone model

Android studio emulator

### Operating system version

Android 6

### Application version and app store

Built from develop (4f74dd4c93b9e1ddd84983b61b73554b74359e71)

### Homeserver

N/A

### Will you send logs?

Yes

### Are you willing to provide a PR?

No

--- a/changelog.d/2846.bugfix
+++ b/changelog.d/2846.bugfix
@@ -1,0 +1,1 @@
+Fix a crash when trying to create an `EncryptedFile` in Android 6.

--- a/libraries/session-storage/impl/src/main/kotlin/io/element/android/libraries/sessionstorage/impl/di/SessionStorageModule.kt
+++ b/libraries/session-storage/impl/src/main/kotlin/io/element/android/libraries/sessionstorage/impl/di/SessionStorageModule.kt
@@ -35,6 +35,13 @@ object SessionStorageModule {
     fun provideMatrixDatabase(@ApplicationContext context: Context): SessionDatabase {
         val name = "session_database"
         val secretFile = context.getDatabasePath("$name.key")
+
+        // Make sure the parent directory of the key file exists, otherwise it will crash in older Android versions
+        val parentDir = secretFile.parentFile
+        if (parentDir != null && !parentDir.exists()) {
+            parentDir.mkdirs()
+        }
+
         val passphraseProvider = RandomSecretPassphraseProvider(context, secretFile)
         val driver = SqlCipherDriverFactory(passphraseProvider)
             .create(SessionDatabase.Schema, "$name.db", context)

